modules/version/src/main/native/android/c/version.c

```diff
@@ -0,0 +1,97 @@
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL GLUON BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#include "util.h"
+
+static jclass jVersionServiceClass;
+static jobject jDalvikVersionService;
+static jmethodID jVersionServiceGetVersion;
+static jmethodID jVersionServiceGetBuild;
+
+static void initializeDalvikHandles() {
+    jVersionServiceClass = GET_REGISTER_DALVIK_CLASS(jVersionServiceClass, "com/gluonhq/helloandroid/DalvikVersionService");
+    ATTACH_DALVIK();
+    jmethodID jVersionServiceInitMethod = (*dalvikEnv)->GetMethodID(dalvikEnv, jVersionServiceClass, "<init>", "(Landroid/app/Activity;)V");
+    jVersionServiceGetVersion = (*dalvikEnv)->GetMethodID(dalvikEnv, jVersionServiceClass, "getVersion", "()Ljava/lang/String;");
+    jVersionServiceGetBuild = (*dalvikEnv)->GetMethodID(dalvikEnv, jVersionServiceClass, "getBuild", "()Ljava/lang/String;");
+
+    jobject jActivity = substrateGetActivity();
+    jobject jtmpobj = (*dalvikEnv)->NewObject(dalvikEnv, jVersionServiceClass, jVersionServiceInitMethod, jActivity);
+    jDalvikVersionService = (*dalvikEnv)->NewGlobalRef(dalvikEnv, jtmpobj);
+    DETACH_DALVIK();
+}
+
+//////////////////////////
+// From Graal to native //
+//////////////////////////
+
+
+JNIEXPORT jint JNICALL
+JNI_OnLoad_version(JavaVM *vm, void *reserved)
+{
+    ATTACH_LOG_INFO("JNI_OnLoad_version called");
+#ifdef JNI_VERSION_1_8
+    JNIEnv* graalEnv;
+    if ((*vm)->GetEnv(vm, (void **)&graalEnv, JNI_VERSION_1_8) != JNI_OK) {
+        ATTACH_LOG_WARNING("Error initializing native Version from OnLoad");
+        return JNI_FALSE;
+    }
+    ATTACH_LOG_FINE("[Version Service] Initializing native Version from OnLoad");
+    initializeDalvikHandles();
+    return JNI_VERSION_1_8;
+#else
+    #error Error: Java 8+ SDK is required to compile Attach
+#endif
+}
+
+// from Java to Android
+
+JNIEXPORT jstring JNICALL Java_com_gluonhq_attach_version_impl_AndroidVersionService_nativeVersion
+(JNIEnv *env, jclass jClass)
+{
+    if (isDebugAttach()) {
+        ATTACH_LOG_FINE("Retrieving Version number\n");
+    }
+    ATTACH_DALVIK();
+    jstring version = (*dalvikEnv)->CallObjectMethod(dalvikEnv, jDalvikVersionService, jVersionServiceGetVersion);
+    const char *responseVersionChars = (*dalvikEnv)->GetStringUTFChars(dalvikEnv, version, 0);
```

Comment:
here and on other places we should later tell dalvikEnv that we don't need access to responseVersionChars anymore. It's tricky, as we might still need it in the regular env. This is probably why we don't have it in most of the services, but it introduces a memory leak.
Would be good to file a separate issue for this, but I'm ok with keeping it like this for now.

Reply:
Filed: https://github.com/gluonhq/attach/issues/281